crypto/serializer/deserializer_lib.c

```diff
@@ -0,0 +1,358 @@
+        return 0;
+    }
+    if ((deser_inst = OPENSSL_zalloc(sizeof(*deser_inst))) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    if (!OSSL_DESERIALIZER_up_ref(deser)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    deser_inst->deser = deser;
+
+    prov = OSSL_DESERIALIZER_provider(deser_inst->deser);
+    provctx = OSSL_PROVIDER_get0_provider_ctx(prov);
+
+    /* Cache the input type for this serializer */
+    params[0] =
+        OSSL_PARAM_construct_utf8_ptr(OSSL_DESERIALIZER_PARAM_INPUT_TYPE,
+                                      (char **)&deser_inst->input_type, 0);
+    params[1] = OSSL_PARAM_construct_end();
+
+    if (!deser_inst->deser->get_params(params)
+        || !OSSL_PARAM_modified(&params[0]))
+        goto err;
+
+    if ((deser_inst->deserctx = deser_inst->deser->newctx(provctx))
+        == NULL)
+        goto err;
+
+    if (sk_OSSL_DESERIALIZER_INSTANCE_push(ctx->deser_insts, deser_inst) <= 0)
+        goto err;
+
+    return 1;
+ err:
+    if (deser_inst != NULL) {
+        if (deser_inst->deser != NULL)
+            deser_inst->deser->freectx(deser_inst->deserctx);
+        OSSL_DESERIALIZER_free(deser_inst->deser);
+        OPENSSL_free(deser_inst);
+    }
+    return 0;
+}
+
+int OSSL_DESERIALIZER_CTX_add_extra(OSSL_DESERIALIZER_CTX *ctx,
+                                    OPENSSL_CTX *libctx, const char *propq)
+{
+    /*
+     * use the input-types of already fetched deserializers to find the next
+     * set of deserializer.  Do this until there are no more to be found.
+     * Do this no more than 10 times.
+     */
+    size_t i = 0, count = 0, depth = 0, end_i = 0;
+
+    if (!ossl_assert(ctx != NULL)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    /*
+     * If there is no stack of OSSL_DESERIALIZER_INSTANCE, we have nothing
+     * more to add.  That's fine.
+     */
+    if (ctx->deser_insts == NULL)
+        return 1;
+
+    end_i = sk_OSSL_DESERIALIZER_INSTANCE_num(ctx->deser_insts);
+    do {
+        for (; i < end_i; i++) {
+            OSSL_DESERIALIZER_INSTANCE *deser_inst =
+                sk_OSSL_DESERIALIZER_INSTANCE_value(ctx->deser_insts, i);
+            const char *name = deser_inst->input_type;
+            OSSL_DESERIALIZER *deser = NULL;
+
+            ERR_set_mark();
+            deser = OSSL_DESERIALIZER_fetch(libctx, name, propq);
+            ERR_pop_to_mark();
+
+            if (deser == NULL)
+                continue;
+
+            /*
+             * We don't care about errors here.  If it doesn't collect,
+             * then it doesn't...
+             */
+            (void)OSSL_DESERIALIZER_CTX_add_deserializer(ctx, deser); /* ref++ */
+            OSSL_DESERIALIZER_free(deser); /* ref-- */
+        }
+        count = sk_OSSL_DESERIALIZER_INSTANCE_num(ctx->deser_insts) - end_i;
```

Comment:
You may want to explain this horribleness here with a comment.

It is adding extra values onto the end of the stack as it tries to iterate over the stack..

Reply:
Why that is horrible, I do not know.  It's pretty damn well controlled, as far as I can see.

Reply:
I've added a huge comment before the start of the outer loop

Reply:
Thinking more about it, I edited this function a bit more to make the sliding windows clearer

Reply:
> Why that is horrible, I do not know. It's pretty damn well controlled, as far as I can see.

Coming from the C++ STL world, adding things to a container generally invalidates iterators.  What does Perl do if you add a new key/value pair inside a foreach loop?  This is probably the first/only occurrence of this kind of thing in openssl.  Adding a comment will help.